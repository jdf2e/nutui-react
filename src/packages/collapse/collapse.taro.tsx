import React, { FunctionComponent, ReactNode, useCallback } from 'react'
import classNames from 'classnames'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'
import CollapseItem from '../collapseitem/index.taro'
import CollapseContext from './context'
import { usePropsValue } from '@/utils/use-props-value'

type keyType = Array<string> | string

export interface CollapseProps extends BasicComponent {
  activeName: keyType
  defaultActiveName: keyType
  accordion: boolean
  expandIcon: ReactNode
  rotate: number
  onChange: (activeName: keyType, name: string, isOpen: boolean) => void
}
const defaultProps = {
  ...ComponentDefaults,
  defaultActiveName: [] as keyType,
  accordion: false,
  expandIcon: null,
  rotate: 180,
} as CollapseProps

export const Collapse: FunctionComponent<Partial<CollapseProps>> & {
  Item: typeof CollapseItem
} = (props) => {
  const {
    className,
    style,
    children,
    activeName,
    defaultActiveName,
    accordion,
    expandIcon,
    rotate,
    onChange,
  } = {
    ...defaultProps,
    ...props,
  }
  const classPrefix = 'nut-collapse'

  const handleChange = (newValue: keyType) => {
    console.log(newValue)
  }

  const [value, setValue] = usePropsValue<keyType>({
    value: activeName,
    defaultValue: defaultActiveName,
    finalValue: [],
    onChange: handleChange,
  })

  const changeVal = (newValue: keyType, name: string, isOpen: boolean) => {
    setValue(newValue)
    onChange && onChange(newValue, name, isOpen)
  }

  const updateValue = (name: string) => {
    if (accordion) {
      if (value === name) {
        changeVal('', name, false)
      } else {
        changeVal(name, name, true)
      }
    } else if (Array.isArray(value)) {
      if (value.includes(name)) {
        const newValue = value.filter((v: string) => v !== name)
        changeVal(newValue, name, false)
      } else {
        const newValue = value.concat([name])
        changeVal(newValue, name, true)
      }
    } else {
      console.warn('[NutUI] <Collapse> 未开启手风琴模式时 activeName 应为数组')
    }
  }

  const isOpen = useCallback(
    (name: string) => {
      if (accordion) {
        return value === name
      }
      if (Array.isArray(value)) {
        return value.includes(name)
      }
      return false
    },
    [accordion, value]
  )

  return (
    <CollapseContext.Provider
      value={{
        isOpen,
        updateValue,
        expandIcon,
        rotate,
      }}
    >
      <div className={classNames(classPrefix, className)} style={style}>
        {children}
      </div>
    </CollapseContext.Provider>
  )
}

Collapse.defaultProps = defaultProps
Collapse.displayName = 'NutCollapse'
Collapse.Item = CollapseItem
