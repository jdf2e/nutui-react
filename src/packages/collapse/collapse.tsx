import React, {
  FunctionComponent,
  useEffect,
  useState,
  memo,
  ReactNode,
} from 'react'
import classNames from 'classnames'

import bem from '@/utils/bem'

export interface CollapseProps {
  className: string
  style: React.CSSProperties
  activeName: Array<number | string> | number | string
  accordion: boolean
  expandIcon: ReactNode
  rotate: number
  onChange: (isOpen: boolean, name: string) => void
  children?: React.ReactNode
}
const defaultProps = {
  style: {},
  activeName: ['0'],
  accordion: false,
  expandIcon: null,
  rotate: 180,
} as CollapseProps

function areEqual(
  prevProps: Partial<CollapseProps>,
  nextProps: Partial<CollapseProps>
) {
  return (
    prevProps.children === nextProps.children &&
    JSON.stringify(prevProps.activeName) ===
      JSON.stringify(nextProps.activeName)
  )
}

export const Collapse: FunctionComponent<Partial<CollapseProps>> = memo(
  (props) => {
    const {
      className,
      style,
      children,
      activeName,
      accordion,
      expandIcon,
      rotate,
      onChange,
    } = {
      ...defaultProps,
      ...props,
    }
    const childrenDom = React.Children.toArray(children)
    const [defaultOpenIndex, setDefaultOpenIndex] = useState<Array<string>>([])
    const handleActiveName = () => {
      let activeArr = []
      if (!Array.isArray(activeName)) {
        activeArr.push(activeName.toString())
      } else {
        // 数组
        if (accordion && activeName.length > 1) {
          console.warn('手风琴模式不支持传递多个打开页签')
        }
        const activeNameStr = activeName.map((item) => {
          return item.toString()
        })
        activeArr = [...activeNameStr]
      }
      return activeArr
    }

    const colBem = bem('collapse')
    useEffect(() => {
      const activeArr = handleActiveName()
      setDefaultOpenIndex(activeArr)
    }, [activeName])

    const onToggle = (isOpen: boolean, name: string) => {
      let newOpenIndex = [...defaultOpenIndex]
      if (isOpen) {
        // 当前状态为true，则变为false,闭合
        const removeIndex = newOpenIndex.findIndex((value) => {
          return value === name
        })
        newOpenIndex.splice(removeIndex, 1)
      } else {
        // 当前状态为false，变为true，展开
        // eslint-disable-next-line no-lonely-if
        if (accordion) {
          newOpenIndex = [name]
        } else {
          newOpenIndex.push(name)
        }
      }
      setDefaultOpenIndex(newOpenIndex)
      onChange && onChange(!isOpen, name)
    }
    return (
      <div className={classNames(colBem(), className)} style={style}>
        {childrenDom.map((item: any) => {
          return React.cloneElement(item, {
            isOpen: defaultOpenIndex.includes(item.props.name),
            onToggle: (isOpen: boolean, name: string) => onToggle(isOpen, name),
            expandIcon: item.props.expandIcon || expandIcon,
            rotate,
            childnull: !!item.props.children,
          })
        })}
      </div>
    )
  },
  areEqual
)

Collapse.defaultProps = defaultProps
Collapse.displayName = 'NutCollapse'
