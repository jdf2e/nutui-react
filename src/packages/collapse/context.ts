import { createContext } from 'react'

const collapseContext = {
  isOpen: (name: string) => {
    return true
  },
  updateValue: (name: string) => {},
  expandIcon: null as any,
  rotate: 180,
}

export default createContext(collapseContext)
