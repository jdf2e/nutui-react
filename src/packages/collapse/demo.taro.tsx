import React, { useState, useEffect } from 'react'
import Taro from '@tarojs/taro'
import { Checked, DownArrow, HeartFill, Star } from '@nutui/icons-react-taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import { Collapse, Button } from '@/packages/nutui.react.taro'
import Header from '@/sites/components/header'

interface itemObj {
  title: string
  name: string
  data: string
}
interface T {
  header1: string
  header2: string
  header3: string
  header4: string
  header5: string
  header6: string
  title1: string
  title2: string
  title3: string
  content1: string
  content2: string
  content3: string
  extra: string
  buttonTextOne: string
  buttonTextSec: string
}
const CollapseDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      header1: '基础用法',
      header2: '无icon样式，绑定点击事件',
      header3: '手风琴模式',
      header4: '自定义折叠图标',
      header5: '自定义 title 与 extra',
      header6: '动态改变数据',
      title1: '标题1',
      title2: '标题2',
      title3: '标题3',
      content1: 'NutUI-React是一套拥有京东风格的轻量级的 React 组件库',
      content2: '在产品的功能、体验、易用性和灵活性等各个方面做了全面的升级！',
      content3: '全面使用 TypeScipt',
      extra: '文本内容',
      buttonTextOne: '改变数据',
      buttonTextSec: '还原数据',
    },
    'zh-TW': {
      header1: '基礎用法',
      header2: '無icon樣式，綁定點擊事件',
      header3: '手風琴模式',
      header4: '自定義折疊圖標',
      header5: '自定義 title & extra',
      header6: '動態改變數據',
      title1: '標題1',
      title2: '標題2',
      title3: '標題3',
      content1: 'Nutui-React 是一套擁有京東風格的輕量級的 React 組件庫',
      content2: '在產品的功能、體驗、易用性和靈活性等各個方面做了全面的升級',
      content3: '全面使用 TypeScipt',
      extra: '文本內容',
      buttonTextOne: '改變數據',
      buttonTextSec: '還原數據',
    },
    'en-US': {
      header1: 'Basic Usage',
      header2: 'No icon style',
      header3: 'accordion Mode',
      header4: 'Custom collapse Icon',
      header5: 'Custom title & extra',
      header6: 'Change Data',
      title1: 'title1',
      title2: 'title2',
      title3: 'title3',
      content1:
        'Nutui-React is a lightweight React component library with JD style',
      content2:
        'The product has been comprehensively upgraded in terms of function, experience, ease of use and flexibility!',
      content3: 'Full use of typescipt',
      extra: 'text content',
      buttonTextOne: 'change data',
      buttonTextSec: 'return data',
    },
  })

  const oldDate = [
    {
      title: translated.title1,
      name: '1',
      data: translated.content1,
    },
    {
      title: translated.title2,
      name: '2',
      data: translated.content1,
    },
    {
      title: translated.title3,
      name: '3',
      data: translated.content1,
    },
  ]

  const newDate = [
    {
      title: `${translated.title1}1`,
      name: '1',
      data: translated.content2,
    },
    {
      title: `${translated.title2}2`,
      name: '2',
      data: translated.content2,
    },
    {
      title: `${translated.title3}3`,
      name: '3',
      data: translated.content2,
    },
  ]

  const [currIndex, setCurrIndex] = useState(2)
  const [domData, setDomData] = useState(oldDate)

  useEffect(() => {
    setDomData(oldDate)
  }, [translated])

  const changeEnv = (isOpen: boolean, name: string) => {
    console.log(isOpen, name)
  }
  const changeNewData = () => {
    setDomData(newDate)
    setCurrIndex(3)
  }
  const changeOldData = () => {
    setDomData(oldDate)
    setCurrIndex(2)
  }
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>translated.header1</h2>
        <Collapse activeKey={['1', '2']} expandIcon={<DownArrow />}>
          <Collapse.Item title={translated.title1} name="1">
            {translated.content1}
          </Collapse.Item>
          <Collapse.Item title={translated.title2} name="2">
            {translated.content2}
          </Collapse.Item>
          <Collapse.Item title={translated.title3} name="3" disabled>
            {translated.content3}
          </Collapse.Item>
        </Collapse>
        <h2>{translated.header2}</h2>
        <Collapse
          activeKey={['1', '2']}
          onChange={(isOpen, name) => changeEnv(isOpen, name)}
        >
          <Collapse.Item title={translated.title1} name="1">
            {translated.content1}
          </Collapse.Item>
          <Collapse.Item title={translated.title2} name="2">
            {translated.content2}
          </Collapse.Item>
          <Collapse.Item title={translated.title3} name="3">
            {translated.content3}
          </Collapse.Item>
        </Collapse>
        <h2>{translated.header3}</h2>
        <Collapse activeKey={['1']} accordion expandIcon={<DownArrow />}>
          <Collapse.Item
            title={translated.title1}
            name="1"
            extra={translated.extra}
          >
            {translated.content1}
          </Collapse.Item>
          <Collapse.Item title={translated.title2} name="2">
            {translated.content2}
          </Collapse.Item>
          <Collapse.Item title={translated.title3} name="3">
            {translated.content3}
          </Collapse.Item>
        </Collapse>
        <h2>{translated.header4}</h2>
        <Collapse
          activeKey={['1']}
          accordion
          expandIcon={<DownArrow />}
          rotate={90}
        >
          <Collapse.Item
            title={translated.title1}
            name="1"
            expandIcon={<Checked />}
          >
            {translated.content1}
          </Collapse.Item>
          <Collapse.Item
            title={translated.title2}
            name="2"
            expandIcon={<HeartFill />}
          >
            {translated.content2}
          </Collapse.Item>
          <Collapse.Item title={translated.title3} name="3">
            {translated.content3}
          </Collapse.Item>
        </Collapse>
        <h2>{translated.header5}</h2>
        <Collapse activeKey={['1']} accordion expandIcon={<Star />}>
          <Collapse.Item
            title={
              <div
                style={{
                  display: 'flex',
                  alignItems: 'center',
                }}
              >
                <Checked />
                {translated.title1}
              </div>
            }
            name="1"
          >
            {translated.content1}
          </Collapse.Item>
          <Collapse.Item
            title={translated.title2}
            extra={
              <div
                style={{
                  display: 'flex',
                  alignItems: 'center',
                }}
              >
                {translated.title2}
                <HeartFill color="red" />
              </div>
            }
            name="2"
          >
            {translated.content2}
          </Collapse.Item>
          <Collapse.Item title={translated.title3} name="3">
            {translated.content3}
          </Collapse.Item>
        </Collapse>

        <h2>{translated.header6}</h2>
        <Collapse activeKey={currIndex} accordion>
          {domData.map((item: itemObj, index: number) => {
            return (
              <Collapse.Item title={item.title} name={item.name} key={index}>
                {item.data}
              </Collapse.Item>
            )
          })}
        </Collapse>
        {/* eslint-disable-next-line react/button-has-type */}
        <Button
          style={{
            marginBottom: '20px',
            marginTop: '20px',
            marginRight: '10px',
          }}
          type="primary"
          size="small"
          onClick={() => changeNewData()}
        >
          {translated.buttonTextOne}
        </Button>
        <Button
          style={{ marginBottom: '20px', marginTop: '20px' }}
          type="info"
          size="small"
          onClick={() => changeOldData()}
        >
          {translated.buttonTextSec}
        </Button>
      </div>
    </>
  )
}

export default CollapseDemo
