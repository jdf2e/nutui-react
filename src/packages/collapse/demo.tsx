import React, { useState } from 'react'
import { Collapse } from './collapse'
import CollapseItem from '../collapseitem'
import { Button } from '../button/button'

const oldDate = [
  {
    title: '标题1',
    name: '1',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
  {
    title: '标题12',
    name: '2',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
  {
    title: '标题13',
    name: '3',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
]
const newDate = [
  {
    title: '标题21',
    name: '1',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
  {
    title: '标题22',
    name: '2',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
  {
    title: '标题23',
    name: '3',
    data: '京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府',
  },
]
const CollapseDemo = () => {
  const [currIndex, setCurrIndex] = useState(2)
  const [domData, setDomData] = useState(oldDate)
  const changeEnv = (isOpen: boolean, name: string) => {
    console.log(isOpen, name)
  }
  const changeNewData = () => {
    setDomData(newDate)
    setCurrIndex(3)
  }
  const changeOldData = () => {
    setDomData(oldDate)
    setCurrIndex(2)
  }
  return (
    <>
      <div className="demo">
        <h2>基础用法</h2>
        <Collapse
          activeName={['1', '2']}
          icon="arrow-down"
          iconSize="16"
          iconColor="#999"
        >
          <CollapseItem title="标题1" name="1">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题2" name="2">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题3" name="3" disabled>
            京东“厂直优品计划”首推“政府优品馆”
          </CollapseItem>
        </Collapse>
        <h2>无icon样式，绑定点击事件</h2>
        <Collapse
          activeName={['1', '2']}
          change={(isOpen, name) => changeEnv(isOpen, name)}
        >
          <CollapseItem title="标题1" name="1">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题2" name="2">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题3" name="3">
            京东“厂直优品计划”首推“政府优品馆”
          </CollapseItem>
        </Collapse>
        <h2>手风琴</h2>
        <Collapse activeName={['1']} accordion icon="arrow-down">
          <CollapseItem title="标题1" name="1" subTitle="文本内容">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题2" name="2">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题3" name="3">
            京东“厂直优品计划”首推“政府优品馆”
          </CollapseItem>
        </Collapse>
        <h2>自定义折叠图标</h2>
        <Collapse activeName={['1']} accordion icon="arrow-right2" rotate={90}>
          <CollapseItem title="标题1" name="1" icon="arrow-down">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题2" name="2" icon="arrow-down">
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题3" name="3" icon="arrow-down">
            京东“厂直优品计划”首推“政府优品馆”
          </CollapseItem>
        </Collapse>
        <h2>自定义title图标</h2>
        <Collapse activeName={['1']} accordion icon="arrow-down">
          <CollapseItem
            title="标题1"
            name="1"
            titleIcon="checked"
            titleIconSize="16"
            titleIconColor="red"
            titleIconPosition="left"
          >
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem
            title="标题2"
            name="2"
            titleIcon="heart-fill"
            titleIconColor="red"
            titleIconPosition="right"
          >
            京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
          </CollapseItem>
          <CollapseItem title="标题3" name="3" icon="arrow-down">
            京东“厂直优品计划”首推“政府优品馆”
          </CollapseItem>
        </Collapse>

        <h2>动态改变数据</h2>
        <Collapse activeName={currIndex} accordion>
          {domData.map((item, index) => {
            return (
              <CollapseItem title={item.title} name={item.name} key={index}>
                {item.data}
              </CollapseItem>
            )
          })}
        </Collapse>
        {/* eslint-disable-next-line react/button-has-type */}
        <Button type="primary" size="small" onClick={() => changeNewData()}>
          改变数据
        </Button>
        <Button type="info" size="small" onClick={() => changeOldData()}>
          还原数据
        </Button>
      </div>
    </>
  )
}

export default CollapseDemo
