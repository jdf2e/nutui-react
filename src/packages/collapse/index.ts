import { Collapse } from './collapse'

export type { CollapseActiveName, CollapseProps } from './collapse'
export default Collapse
