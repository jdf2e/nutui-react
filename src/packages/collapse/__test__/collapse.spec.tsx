import * as React from 'react'
// import * as renderer from 'react-test-renderer'

import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { ArrowDown, ArrowRight } from '@nutui/icons-react'
import { Collapse } from '../collapse'
import { CollapseItem } from '../../collapseitem/collapseitem'

test('should match snapshot', () => {
  const { asFragment } = render(
    <>
      <Collapse activeName={['1', '2']} expandIcon={<ArrowDown />}>
        <CollapseItem title="标题1" name="1">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" disabled>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  expect(asFragment()).toMatchSnapshot()
})

test('prop icon iconSize iconColor', () => {
  const { getByTestId, container } = render(
    <>
      <Collapse activeName={['1', '2']} expandIcon={<ArrowDown />}>
        <CollapseItem title="标题1" name="1">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" disabled>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  expect(
    container.querySelector('.nut-collapse-item-icon .nut-icon')
  ).toHaveClass('nut-icon-ArrowDown')
})

test('prop activeName', () => {
  const { container } = render(
    <>
      <Collapse activeName={['1', '2']} expandIcon={<ArrowDown />}>
        <CollapseItem title="标题1" name="1">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" disabled>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )

  setTimeout(() => {
    expect(container.querySelector('.nut-collapse-item-content')).toHaveStyle(
      'height: 66px'
    )
  }, 100)
})

test('prop rotate', () => {
  const { getByTestId, container } = render(
    <>
      <Collapse
        activeName={['1']}
        accordion
        expandIcon={<ArrowRight />}
        rotate={180}
      >
        <CollapseItem
          title="标题1"
          name="1"
          expandIcon={<ArrowDown />}
          data-testid="collapse-one"
        >
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2" expandIcon={<ArrowDown />}>
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" expandIcon={<ArrowDown />}>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  fireEvent.click(getByTestId('collapse-one'))
  setTimeout(() => {
    expect(
      getByTestId('collapse-one').querySelector('.nut-collapse-item-icon')
    ).toHaveStyle('transform: translateY(-50%) rotate(180deg);')
  }, 100)
})

test('prop title extra', () => {
  const { getByTestId, container } = render(
    <>
      <Collapse activeName={['1']} accordion expandIcon={<ArrowRight />}>
        <CollapseItem
          title="标题1"
          extra="副标题"
          name="1"
          data-testid="collapse-one"
        >
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2" expandIcon={<ArrowDown />}>
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" expandIcon={<ArrowDown />}>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  expect(
    getByTestId('collapse-one').querySelector('.nut-collapse-item-title')
  ).toHaveTextContent('标题1')
  expect(
    getByTestId('collapse-one').querySelector('.nut-collapse-item-extra')
  ).toHaveTextContent('副标题')
  expect(getByTestId('collapse-one').querySelector('.nut-icon')).toHaveClass(
    'nut-icon-ArrowRight'
  )
})

test('event onChange & prop disabled', () => {
  const handleChange = jest.fn()
  const { getByTestId } = render(
    <>
      <Collapse defaultActiveName={['1']} onChange={handleChange}>
        <CollapseItem title="标题1" name="1" data-testid="item1">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2" data-testid="item2">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" data-testid="item3" disabled>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  const item1 = getByTestId('item1').querySelector('.nut-collapse-item-title')
  const item2 = getByTestId('item2').querySelector('.nut-collapse-item-title')
  const item3 = getByTestId('item3').querySelector('.nut-collapse-item-title')
  fireEvent.click(item3 as Element)
  expect(handleChange).not.toBeCalled()
  fireEvent.click(item2 as Element)
  expect(handleChange).toBeCalledWith(['1', '2'], '2', true)
  setTimeout(() => {
    fireEvent.click(item1 as Element)
    expect(handleChange).toBeCalledWith(
      ['1', '2'],
      '2',
      true,
      ['2'],
      '1',
      false
    )
  }, 0)
})

test('event onChange & prop accordion', () => {
  const handleChange = jest.fn()
  const { getByTestId } = render(
    <>
      <Collapse defaultActiveName="1" accordion onChange={handleChange}>
        <CollapseItem title="标题1" name="1" data-testid="item1">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题2" name="2" data-testid="item2">
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </CollapseItem>
        <CollapseItem title="标题3" name="3" data-testid="item3" disabled>
          京东“厂直优品计划”首推“政府优品馆”
        </CollapseItem>
      </Collapse>
    </>
  )
  const item2 = getByTestId('item2').querySelector('.nut-collapse-item-title')
  const item3 = getByTestId('item3').querySelector('.nut-collapse-item-title')
  fireEvent.click(item3 as Element)
  expect(handleChange).not.toBeCalled()
  fireEvent.click(item2 as Element)
  expect(handleChange).toBeCalledWith('2', '2', true)
  setTimeout(() => {
    fireEvent.click(item2 as Element)
    expect(handleChange).toBeCalledWith('2', '2', true, '', '2', false)
  }, 0)
})
