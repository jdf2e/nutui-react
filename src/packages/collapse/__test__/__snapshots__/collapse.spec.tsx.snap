// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="nut-collapse"
  >
    <div
      class="nut-collapse-item"
    >
      <div
        class="nut-collapse-item__header"
      >
        <div
          class="nut-collapse-item__title"
        >
          标题1
        </div>
        <div
          class="nut-collapse-item__sub-title"
        />
        <div
          class="nut-collapse-item__icon-box"
        >
          <div
            class="nut-collapse-item__icon"
            style="transform: translateY(-50%) rotate(180deg);"
          >
            <i
              class=" nut-icon nutui-iconfont nut-icon-arrow-down "
              style="color: rgb(153, 153, 153); font-size: 16px; width: 16px; height: 16px;"
            />
          </div>
        </div>
      </div>
      <div
        class="nut-collapse-item__content"
        style="height: 0px;"
      >
        <div
          class="nut-collapse-item__content-text"
        >
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </div>
      </div>
    </div>
    <div
      class="nut-collapse-item"
    >
      <div
        class="nut-collapse-item__header"
      >
        <div
          class="nut-collapse-item__title"
        >
          标题2
        </div>
        <div
          class="nut-collapse-item__sub-title"
        />
        <div
          class="nut-collapse-item__icon-box"
        >
          <div
            class="nut-collapse-item__icon"
            style="transform: translateY(-50%) rotate(180deg);"
          >
            <i
              class=" nut-icon nutui-iconfont nut-icon-arrow-down "
              style="color: rgb(153, 153, 153); font-size: 16px; width: 16px; height: 16px;"
            />
          </div>
        </div>
      </div>
      <div
        class="nut-collapse-item__content"
        style="height: 0px;"
      >
        <div
          class="nut-collapse-item__content-text"
        >
          京东“厂直优品计划”首推“政府优品馆” 3年覆盖80%镇级政府
        </div>
      </div>
    </div>
    <div
      class="nut-collapse-item"
    >
      <div
        class="nut-collapse-item__header nut-collapse-item__header--disabled"
      >
        <div
          class="nut-collapse-item__title"
        >
          标题3
        </div>
        <div
          class="nut-collapse-item__sub-title"
        />
        <div
          class="nut-collapse-item__icon-box"
        >
          <div
            class="nut-collapse-item__icon"
            style="transform: translateY(-50%);"
          >
            <i
              class=" nut-icon nutui-iconfont nut-icon-arrow-down "
              style="color: rgb(194, 194, 194); font-size: 16px; width: 16px; height: 16px;"
            />
          </div>
        </div>
      </div>
      <div
        class="nut-collapse-item__content"
        style="height: 0px;"
      >
        <div
          class="nut-collapse-item__content-text"
        >
          京东“厂直优品计划”首推“政府优品馆”
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
